File: AidlServicePoC/app/src/main/java/com/asfartz/aidlservicepoc/MathService.java
package com.asfartz.aidlservicepoc;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.os.RemoteException;
import android.util.Log;

public class MathService extends Service {

    private static final String TAG = "Andy";

    public MathService() {
    }

    @Override
    public IBinder onBind(Intent intent) {
        Log.d(TAG, "onBind: intent = " + intent);
        Log.d(TAG, "onBind: returning mBinder");
        return mBinder;
    }

    IMathService.Stub mBinder = new IMathService.Stub() {
        @Override
        public int add(int x, int y) throws RemoteException {
            return x + y;
        }

        @Override
        public int substract(int x, int y) throws RemoteException {
            return x - y;
        }
    };

}

File: ServerApp/app/src/main/java/com/ansfartz/serverapp/MathService.kt
package com.ansfartz.serverapp

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.util.Log

/**
 * Service class which makes the calls to the implementation of the AIDL interface [MathManagerImpl]
 */
class MathService : Service() {

    private val mathManagerImpl = MathManagerImpl()

    override fun onCreate() {
        Log.d("MathService", "onCreate:")
        super.onCreate()
    }

    override fun onBind(intent: Intent?): IBinder {
        Log.d("MathService", "onBind: intent: getPackage = ${intent!!.`package`} " +
                "getAction = ${intent.action} " +
                "getData = ${intent.data} " +
                "getComponent = ${intent.component} " +
                "getScheme = ${intent.scheme} " +
                "getDataString = ${intent.dataString}"
        )

        return mathManagerImpl
    }

    override fun onDestroy() {
        Log.d("MathService", "onCreate:")
        super.onDestroy()
    }
}